text/0000-crt-static.md

```diff
@@ -0,0 +1,371 @@
+  be a long time before Cargo itself can rebuild std.
+
+### Specifying dynamic/static C runtime linkage
+
+A new `target-feature` flag will now be supported by the compiler for relevant
+targets: `crt-static`. This can be enabled and disabled in the compiler via:
+
+```
+rustc -C target-feature=+crt-static ...
+rustc -C target-feature=-crt-static ...
+```
+
+Currently all `target-feature` flags are passed through straight to LLVM, but
+this proposes extending the meaning of `target-feature` to Rust-target-specific
+features as well. Target specifications will be able to indicate what custom
+target-features can be defined, and most existing targets will define a new
+`crt-static` feature which is turned off by default (except for musl).
+
+The default of `crt-static` will be different depending on the target. For
+example `x86_64-unknown-linux-musl` will have it on by default, whereas
+`arm-unknown-linux-musleabi` will have it turned off by default.
+
+### Lowering `cfg` values to Cargo build script environment variables
+
+Cargo will begin to forward `cfg` values from the compiler into build
+scripts. Currently the compiler supports `--print cfg` as a flag to print out
+internal cfg directives, which Cargo uses to implement platform-specific
+dependencies.
+
+When Cargo runs a build script it already sets a [number of environment
+variables][cargo-build-env], and it will now set a family of `CARGO_CFG_*`
+environment variables as well. For each key printed out from `rustc --print
+cfg`, Cargo will set an environment variable for the build script to learn
+about.
+
+[cargo-build-env]: http://doc.crates.io/environment-variables.html#environment-variables-cargo-sets-for-build-scripts
+
+For example, locally `rustc --print cfg` prints:
+
+```
+target_os="linux"
+target_family="unix"
+target_arch="x86_64"
+target_endian="little"
+target_pointer_width="64"
+target_env="gnu"
+unix
+debug_assertions
+```
+
+And with this Cargo would set the following environment variables for build
+script invocations for this target.
+
+```
+export CARGO_CFG_TARGET_OS=linux
+export CARGO_CFG_TARGET_FAMILY=unix
+export CARGO_CFG_TARGET_ARCH=x86_64
+export CARGO_CFG_TARGET_ENDIAN=little
+export CARGO_CFG_TARGET_POINTER_WIDTH=64
+export CARGO_CFG_TARGET_ENV=gnu
+export CARGO_CFG_UNIX
+export CARGO_CFG_DEBUG_ASSERTIONS
+```
+
+As mentioned in the previous section, the linkage of the C standard library will
+be specified as a target feature, which is lowered to a `cfg` value, thus giving
+build scripts the ability to modify compilation options based on C standard
+library linkage. One important complication here is that `cfg` values in Rust
+may be defined multiple times, and this is the case with target features. When a
+`cfg` value is defined multiple times, Cargo will create a single environment
+variable with a comma-separated list of values.
+
+So for a target with the following features enabled
+
+```
+target_feature="sse"
```

Comment:
Will Cargo set the `CARGO_CFG_TARGET_FEATURE` variable for CPU features that are *implied* by the
"llvm-target" field of a target specification? For example:

Will this

`cargo build --target x86_64-unknown-linux-gnu`

produce this

```
CARGO_CFG_TARGET_FEATURE="mmx,sse,sse2,."
```

?

And this

`RUSTFLAGS="-mmx,-sse,-sse2" cargo build --target x86_64-unknown-linux-gnu`

produce this

```
CARGO_CFG_TARGET_FEATURE=""
```

?

Because, otherwise, this scheme would leave the `CARGO_CFG_TARGET_FEATURE` empty for both `cargo
build` cases and that's not an accurate representation of the target CPU features.

And if Cargo will, indeed, set these implicit CPU features based on the "llvm-target", how will that
be implemented? Will Cargo call ("shell out to") a new `rustc --print llvm-target-features` command
created for this purpose?